### PR TITLE
Companion for substrate #7878

### PR DESCRIPTION
**Pull request by drahnr**
Improves error handling in polkadot keeping the `source` alive by removing all the silly `Error -> String -> Error` transitions.

Ref https://github.com/paritytech/substrate/pull/7446 #1863

### CODE REVIEW

File: parachain/test-parachains/adder/collator/src/main.rs [hunk 1]
```diff
@@ -37,7 +38,7 @@ fn main() -> Result<()> {
 			let collator = Collator::new();
 			println!("0x{:?}", HexDisplay::from(&collator.genesis_head()));
 
-			Ok(())
+			Ok::<_, Error>(())
```
Comment (ordian) [inline parachain/test-parachains/adder/collator/src/main.rs:41]: I wonder why this type annotation is required here, but not below in `cli::Subcommand::ExportGenesisWasm` branch, and why at all.

**Review by ordian — APPROVED** [review]
The companion update looks straightforward, the only worry is how will this affect compile times due to generics use (and maybe even binary size).

File: cli/src/command.rs [hunk 2]
```diff
@@ -168,34 +168,36 @@ pub fn run() -> Result<()> {
...
 
 			runner.async_run(|mut config| {
-				let (client, _, import_queue, task_manager) = service::new_chain_ops(&mut config, None)
-					.map_err(|e| sc_service::Error::Other(e.to_string()))?;
-				Ok((cmd.run(client, import_queue), task_manager))
+				let (client, _, import_queue, task_manager) = service::new_chain_ops(&mut config, None)?;
+				Ok((async move {
+					cmd.run(client, import_queue).await
+						.map_err(|e| Error::SubstrateCli(e))
+				}, task_manager))
```
Comment (bkchr) [inline cli/src/command.rs:200]: Why not https://docs.rs/futures/0.3.9/futures/future/trait.TryFutureExt.html#method.map_err?

I don't see any reason to use await here.

File: parachain/test-parachains/adder/collator/src/main.rs (hunk 1 above)
Comment (drahnr) [reply parachain/test-parachains/adder/collator/src/main.rs:41]: The return type could technically by anything that impl `From<..>` error and that way it's tied down in the first branch.

File: cli/src/command.rs (hunk 2 above)
Comment (drahnr) [reply cli/src/command.rs:200]: That was my first approach, but it yielded resolution errors when importing `use futures::future::FutureExt` with no added benefit. Is there a document outlining perf delta between an `async move {}` block and future combinators?

**Review by coriolinus — APPROVED** [review]
This is a good idea.

**Pull request closed without merging**